deeppavlov/dataset_iterators/multitask_pal_bert_iterator.py

```diff
@@ -0,0 +1,322 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import math
+import numpy as np
+from logging import getLogger
+from typing import Iterator, List, Optional, Tuple, Union, Dict
+
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.common.params import from_params
+from deeppavlov.core.data.data_learning_iterator import DataLearningIterator
+
+log = getLogger(__name__)
+
+
+@register('multitask_pal_bert_iterator')
+class MultiTaskPalBertIterator:
+    """
+    Class merges data from several dataset iterators. When used for batch generation batches from
+    merged dataset iterators are united into one batch. If sizes of merged datasets are different
+    smaller datasets are repeated until their size becomes equal to the largest dataset.
+
+    Args:
+        data: dictionary which keys are task names and values are dictionaries with fields
+            ``"train", "valid", "test"``.
+        tasks: dictionary which keys are task names and values are init params of dataset iterators.
+
+    Attributes:
+        data: dictionary of data with fields "train", "valid" and "test" (or some of them)
+    """
+
+    def __init__(
+        self,
+        data: dict,
+        num_train_epochs: int,
+        steps_per_epoch: int,
+        tasks: dict,
+        gradient_accumulation_steps: Optional[int] = 1,
+    ):
+        self.task_iterators = {}
+        for task_name, task_iterator_params in tasks.items():
+            task_iterator_params = copy.deepcopy(task_iterator_params)
+            task_iterator_params["class_name"] = task_iterator_params[
+                "iterator_class_name"
+            ]
+            del task_iterator_params["iterator_class_name"]
+            self.task_iterators[task_name] = from_params(
+                task_iterator_params, data=data[task_name]
+            )
+        self.n_tasks = len(tasks.keys())
+        self.num_train_epochs = num_train_epochs
+        self.steps_per_epoch = steps_per_epoch
+        self.gradient_accumulation_steps = gradient_accumulation_steps
+        self.epochs_done = 0
+        self.steps_taken = 0
+        self.task_id = None
+        self.data = {
+            "train": self._extract_data_type("train"),
+            "valid": self._extract_data_type("valid"),
+            "test": self._extract_data_type("test"),
+            "all": self._unite_dataset_parts(
+                self._extract_data_type("train"),
```

Comment:
we are calling twice self._extract_data_type("train") , self._extract_data_type("valid"), self._extract_data_type("test")

Reply:
Fixed